plugin/trino-http-event-listener/src/main/java/io/trino/plugin/httpquery/HttpEventListener.java

```diff
@@ -106,22 +108,30 @@ public void splitCompleted(SplitCompletedEvent splitCompletedEvent)
                 .setBodyGenerator(out -> objectWriter.writeValue(out, event))
                 .build();
 
-        attemptToSend(request, 0);
+        attemptToSend(request, 0, this.config.getRetryDelay());
     }
 
-    private void attemptToSend(Request request, int attempt)
+    private void attemptToSend(Request request, int attempt, Duration delay)
     {
         Futures.addCallback(client.executeAsync(request, createStatusResponseHandler()), new FutureCallback<>()
         {
             @Override
             public void onSuccess(StatusResponse result)
             {
-                if (result != null && result.getStatusCode() >= 500 && attempt < config.getRetryCount()) {
-                    attemptToSend(request, attempt + 1);
+                verify(result != null);
+
+                if (result.getStatusCode() >= 500 && attempt < config.getRetryCount()) {
+                    try {
+                        Thread.sleep(delay.roundTo(TimeUnit.SECONDS));
```

Comment:
with `directExecutor` we may end up block main thread which is calling the event listener.
There is a race condition between HTTP client thread which executes request and main thread which is calling `Futures.addCallback`.
If the `Futures.addCallback` is executed after HTTP request is already completed then the main thread will get blocked for arbitrary amount of time.

As an alternative you may have `ScheduledExecutorService` in `HttpEventListener`. Then you can use [schedule()](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ScheduledExecutorService.html#schedule(java.util.concurrent.Callable,%20long,%20java.util.concurrent.TimeUnit)) to schedule tasks which will perform actual request scheduling in httpClient.
The `ScheduledExecutorService` may be backed by single threaded executor as it will never do any work which may block.

Reply:
That is a very good point, `directExecutor` has been there since before retries were implemented and in that scenario, it didn't pose a problem since the callback was light. 

The docstring from Guava actually warns about this exact case:
```
A call like future.transform(function, directExecutor()) may execute the function immediately in the thread that is calling transform. (This specific case happens if the future is already completed.) 
````